### PR TITLE
Search sharees on lookup server when explicitly requested by user

### PR DESCRIPTION
**Pull request by ChristophWurst**
# Before

Sharee search either never or always goes to lookup server, depending on admin setting.

# Now

Sharee search by default just searches locally. If no results are returned and the lookup server is not explicitly disabled by the admin, the user gets the option to *search globally*. This retriggers the search request but also includes lookup server results.

---

Note: I had some difficulties connection to the lookup server form my dev server but the default empty response for that is sufficient for testing :)

---

Todo

- [x] Design does not look like @jancborchardt's mockup (search icon has to be moved *before* the text)

### CODE REVIEW

**Comment by jancborchardt** [comment]
For the record here’s the mockup :) 
search mockup


Btw @ChristophWurst, seems some unrelated files were edited maybe in the fixup?

> #### Conflicting files
> `apps/accessibility/js/accessibility.js`
> `apps/accessibility/js/accessibility.js.map`

**Comment by ChristophWurst** [comment]
> Btw @ChristophWurst, seems some unrelated files were edited maybe in the fixup?

Should be good now.

Could you please test?

**Comment by jancborchardt** [comment]
> Note: I had some difficulties connection to the lookup server form my dev server but the default empty response for that is sufficient for testing :)

So I need to set up something specific for testing? Because right now I don’t get anything displayed additionally in the suggestions.

**Comment by ChristophWurst** [comment]
> So I need to set up something specific for testing? Because right now I don’t get anything displayed additionally in the suggestions.

Yeah, there seem to be hiccups with the lookup server according to @schiessle. I just tested with the default error handling of no search results. But the switch to global search does work, right?

**Comment by jancborchardt** [comment]
> But the switch to global search does work, right?

No, putting in something into the search field only shows the suggested people from my server, no entry to "Search globally" or similar.

**Comment by ChristophWurst** [comment]
> No, putting in something into the search field only shows the suggested people from my server, no entry to "Search globally" or similar.

Are on https://github.com/nextcloud/server/pull/14180? Because this branch here should not show you suggestions/recommendations.

Note that it currently does only give you the option to search globally if no local search results are found. Should be be shown regardless of that?

**Comment by ChristophWurst** [comment]
> Should be be shown regardless of that?

One problem that arises with that is that if we show the "Search globally" last, it won't be visible with many local search results unless we put it as first element.

You were suggesting something like a sticky position. I'm afraid it would be very hard to do so with this legacy jquery UI component.

**Comment by jancborchardt** [comment]
> Because this branch here should not show you suggestions/recommendations.

Yep, that’s why I said "putting in something into the search field" :) I am on this branch.

> Note that it currently does only give you the option to search globally if no local search results are found. Should be be shown regardless of that?

Yes, should be shown regardless of that as per mockup. :)



> One problem that arises with that is that if we show the "Search globally" last, it won't be visible with many local search results unless we put it as first element.
> You were suggesting something like a sticky position. I'm afraid it would be very hard to do so with this legacy jquery UI component.

Yep, so we have 3 options for the posiitoning of the "Search globally":
1. Show it last with a sticky position. Could be hard to implement wiht the jQuery component as you said – any idea @nextcloud/designers?
2. Show it last without a specific position. Good intermediate step towards 1 anyway so let’s do that as a start?
3. Show it in _first_ position. If we do this, then it will mess with the "press Enter to select first result" and it will be _way_ too present, so I’d rather avoid this.

**Comment by ChristophWurst** [comment]
> Show it last without a specific position. Good intermediate step towards 1 anyway so let’s do that as a start?

Yep, done :+1:

**Comment by ChristophWurst** [comment]
There seem to be a few UI hiccups here and there (already seen when working on #14180, hence this might be an old bug) but this generally works and also looks like @jancborchardt's mockup.

**Comment by schiessle** [comment]
I just tried it. But I always get a "no user found" and a endless spinner. (Some side node: I can't directly retry it but always have to reload the page to get the 'seach globally' option back).

The lookup server seems to work fine in my test setup and return following object if I search for "Max":

````
[{"federationId":"admin@localhost\\/server","name":{"value":"Max Mustermann","verified":"1"},"email":{"value":"bjoern@schiessle.org","verified":"1"}}]
````

**Comment by ChristophWurst** [comment]
> I just tried it. But I always get a "no user found" and a endless spinner. (Some side node: I can't directly retry it but always have to reload the page to get the 'seach globally' option back).

Yeah, because it remembers your choice :wink: Should it return to local search after showing the global results? I thought this would be a bit weird so I made it remember it for the current file/folder.

> The lookup server seems to work fine in my test setup and return following object if I search for "Max":

Can you inspect the HTTP response from the sharee search in the browser? If that's empty as well, could you attach a debugger and see if the lookup server integration passes on that data? I couldn't get the lookup server to work.

**Comment by schiessle** [comment]
HTTP response is 200:

````
{"ocs":{"meta":{"status":"ok","statuscode":100,"message":"OK","totalitems":"","itemsperpage":""},"data":{"exact":{"users":[],"groups":[],"remotes":[],"remote_groups":[],"emails":[],"circles":[],"rooms":[],"lookup":[]},"users":[],"groups":[],"remotes":[],"remote_groups":[],"emails":[],"lookup":[],"circles":[],"rooms":[],"lookupEnabled":true}}}
````

And as said, the lookup server returns a valid result? So maybe we fail to detect/read the result from the lookup server correctly? Do we filter out some of the results? Because the user in my test case is the same user who performs the search.

**Comment by ChristophWurst** [comment]
> HTTP response is 200:

`"lookup":[]` tells me the search back-end does not pass on that data. I assume that's also an issue on master but I'll look into it!

**Comment by ChristophWurst** [comment]
Turns out @schiessle was logged in as the user that should be shown, but we filter that: https://github.com/nextcloud/server/blob/4c4f798bac7110bc38ecbbdc20eb690c2742dad4/lib/private/Collaboration/Collaborators/LookupPlugin.php#L100 :wink:

**Comment by ChristophWurst** [comment]
- [ ] BUG: after clicking *Search globally* and fetching the data, the list does not show automatically but I have to leave the input and focus again

**Review by schiessle — APPROVED** [review]
Tested and works great! :+1:

**Comment by ChristophWurst** [comment]
Thanks for testing! :raised_hands: 

I squashed my fixup commits and rebased onto master.

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

File: core/js/sharedialogview.js [hunk 1]
```diff
@@ -739,15 +747,20 @@
 					}
 
 				} else {
-					var title = t('core', 'No users or groups found for {search}', {search: $shareWithField.val()});
+					var title;
+					if (view._lookup) {
+						title = t('core', 'No (global) users found for {search}', {search: $shareWithField.val()});
```
Comment (MorrisJobke) [inline core/js/sharedialogview.js]: I was a bit confused by the "(global)" in parentheses. Should we really add the parentheses there or just drop then to have a proper sentence shown in the UI?

**Comment by MorrisJobke** [comment]
CI is failing a lot - please have a look.

**Comment by rullzer** [comment]
Yes seems the tests need an update (as well as extention)

File: core/js/sharedialogview.js (hunk 1 above)
Comment (rullzer) [reply core/js/sharedialogview.js]: I guess the default is still good enough indeed. I guess users do not care if it is global or not.

**Comment by MorrisJobke** [comment]
Still failing tests

**Comment by ChristophWurst** [comment]
In my defense: they pass locally

```
$ NOCOVERAGE=1 ./autotest.sh sqlite apps/files_sharing/tests/Controller/ShareesAPIControllerTest.php
Using PHP executable /usr/bin/php
Parsing all files in lib/public for the presence of @since or @deprecated on each method...

Using database oc_autotest
Setup environment for sqlite testing on local storage ...
Installing ....
Nextcloud was successfully installed
Testing with sqlite ...
No coverage
/home/christoph/.composer/vendor/bin/phpunit --configuration phpunit-autotest.xml --log-junit autotest-results-sqlite.xml ../apps/files_sharing/tests/Controller/ShareesAPIControllerTest.php 
PHPUnit 6.5.7 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.3.2 with Xdebug 2.7.0RC2
Configuration: /home/christoph/workspace/nextcloud/tests/phpunit-autotest.xml

............................................                      44 / 44 (100%)

Time: 1.61 seconds, Memory: 32.00MB

OK (44 tests, 212 assertions)
```

On CI
```
1) OCA\Files_Sharing\Tests\Controller\ShareesAPIControllerTest::testSearch with data set #0 (array(), '', 'yes', true, true, true, array(0, 1, 4), false, true, true)
OCP\IConfig::getAppValue('files_sharing', 'lookupServerEnabled', 'yes') was not expected to be called more than 2 times.

/drone/src/github.com/nextcloud/server/apps/files_sharing/lib/Controller/ShareesAPIController.php:216
/drone/src/github.com/nextcloud/server/apps/files_sharing/tests/Controller/ShareesAPIControllerTest.php:299
```

File: core/js/sharedialogview.js (hunk 1 above)
Comment (ChristophWurst) [reply core/js/sharedialogview.js]: Yeah, will revert this change.

**Comment by ChristophWurst** [comment]
Pushed a fixup to fix *some* of the jsunit failures (the other CI failures are unrelated from what I can tell). Let's see what else fails. I can't really get the js tests to pass locally, so this is a bit of trial and error ...

**Comment by MorrisJobke** [comment]
> Pushed a fixup to fix _some_ of the jsunit failures (the other CI failures are unrelated from what I can tell). Let's see what else fails. I can't really get the js tests to pass locally, so this is a bit of trial and error ...

Webpack and JSUnit ;)

**Comment by ChristophWurst** [comment]
> Webpack and JSUnit ;)

Yeah. Found out that some of the shipped apps still had their own `node_modules` and therefore I produced weird bundles. That is fixed.

Also, jsunit *should* be happy now.

**Comment by ChristophWurst** [comment]
Aaaaaaah now CI looks good :) Let me do a final squash&rebase

**Pull request merged**